src/wasm/asm-wasm-builder.cc

```diff
@@ -355,21 +376,22 @@ class AsmWasmBuilderImpl : public AstVisitor {
   }
 
   void VisitAssignment(Assignment* expr) {
-    DCHECK(in_function_);
-    BinaryOperation* value_op = expr->value()->AsBinaryOperation();
-    if (value_op != NULL && MatchBinaryOperation(value_op) == kAsIs) {
-      VariableProxy* target_var = expr->target()->AsVariableProxy();
-      VariableProxy* effective_value_var = GetLeft(value_op)->AsVariableProxy();
-      if (target_var != NULL && effective_value_var != NULL &&
-          target_var->var() == effective_value_var->var()) {
-        block_size_--;
-        return;
+    if (in_function_) {
+      BinaryOperation* value_op = expr->value()->AsBinaryOperation();
+      if (value_op != NULL && MatchBinaryOperation(value_op) == kAsIs) {
+        VariableProxy* target_var = expr->target()->AsVariableProxy();
+        VariableProxy* effective_value_var = GetLeft(value_op)->AsVariableProxy();
+        if (target_var != NULL && effective_value_var != NULL &&
+            target_var->var() == effective_value_var->var()) {
+          block_size_--;
+          return;
```

Comment:
I am not sure if that's a good idea. The whole reason to even do this is to reduce the size of the binary. Right now we are removing 3 bytes for every param (I think that is the only significant use case where this optimization gives any benefit). If we move to NOP, we will be removing 2 instead. However, with all the other possible inefficiencies in there, I am not sure how much this optimization is worth. Will let you make a final call